compiler/src/dotty/tools/dotc/transform/TypeTestsCasts.scala

```diff
@@ -153,6 +153,33 @@ object TypeTestsCasts {
               derivedTree(e, defn.Any_isInstanceOf, e.tpe)
                 .and(isArrayTest(e))
             }
+          case arg if arg.isRef(defn.SingletonClass) =>
+            // Does the expression type contain a union type?
+            var containsUnion: Boolean = false
+
+            def isSingleton(tp: Type): Boolean = tp.dealias match {
+              case tp: SingletonType =>
+                true
+              case AndType(tp1, tp2) =>
+                isSingleton(tp1) || isSingleton(tp2)
+              case tp =>
+                if (tp.isInstanceOf[OrType])
```

Comment:
Why not simply:

    caes OrType(tp1, tp2) => isSingleton(tp1) && isSingleton(tp2)

(with a comment that this case currently cannot arise, but is left there anyway for future-proofing)?

Reply:
The union of two singleton types might not be a singleton type, e.g. "1 | 2" contains two values, not one.